src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+                       struct ImageInfo* info)
+{
+    try {
+        CHECK(self_);
+        ((struct SpinnakerCamera*)self_)->get_frame(im, nbytes, info);
+        return Device_Ok;
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+    return Device_Err;
+}
+
+//
+// Camera implementation
+//
+
+SpinnakerCamera::SpinnakerCamera(Spinnaker::CameraPtr camera)
+  : Camera{ .set = ::spinnakercam_set,
+            .get = ::spinnakercam_get,
+            .get_meta = ::spinnakercam_get_meta,
+            .get_shape = ::spinnakercam_get_shape,
+            .start = ::spinnakercam_start,
+            .stop = ::spinnakercam_stop,
+            .execute_trigger = ::spinnakercam_execute_trigger,
+            .get_frame = ::spinnakercam_get_frame,
+  }
+  , camera_(camera)
+  , last_known_settings_{}
+  , frame_id_(0)
+  , started_(false)
+{
+    // Sometimes the camera is still initialized from a previous run.
+    // Ideally, we would error here, but that can make the camera indefinitely
+    // unusable in Acquire, so log and reset instead.
+    if (camera->IsInitialized()) {
+        LOGE("Camera was already initialized. De-initializing it before "
+             "re-initializing.");
+        camera_->DeInit();
+    }
+    camera->Init();
+
+    // Acquire only supports certain values of some node values, so set these
+    // once on initialization before getting or setting any other node values
+    // which may depend on them.
+    set_enum_node(camera_->ExposureAuto, genicam_off);
+    set_enum_node(camera_->ExposureMode, genicam_timed);
+
+    get(&last_known_settings_);
+}
+
+SpinnakerCamera::~SpinnakerCamera()
+{
+    try {
+        stop();
+        const std::scoped_lock lock(lock_);
+        camera_->DeInit();
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+}
+
+void
+SpinnakerCamera::set(struct CameraProperties* properties)
+{
+    const std::scoped_lock lock(lock_);
+    // Set shape before offset because Spinnaker blocks updates
+    // to offset if the shape is too big.
+    maybe_set_shape(properties->shape);
+    maybe_set_offset(properties->offset);
+    maybe_set_binning(properties->binning);
+    maybe_set_sample_type(properties->pixel_type);
+    // Exposure time can depend on lots of properties, so update it last.
+    maybe_set_exposure_time_us(properties->exposure_time_us);
+    maybe_set_input_trigger_frame_start(properties->input_triggers.frame_start);
+    maybe_set_output_trigger_exposure(properties->output_triggers.exposure);
+}
+
+void
+SpinnakerCamera::maybe_set_exposure_time_us(float target_us)
+{
+    if (target_us != last_known_settings_.exposure_time_us) {
+        set_float_node(camera_->ExposureTime, (double)target_us);
+        // TODO: may not need to actually get from camera because try_camera_set
```

Comment:
`last_known_settings_` might be used in two ways. First, to quickly return the last known settings via get. Second, as a memoizing strategy to reduce the number of Spinnaker calls make on `set`.

I think that Second use case is the important one. And so that would be why you might want to be sure to update it here.

Reply:
The problem with both uses is that we're assuming that Acquire is the only thing modifying the camera settings while its process is alive. That might be true in most cases, but it's not guaranteed to be true (e.g. I think I can successfully modify settings with SpinView while Acquire is running).

Reply:
For now, I've removed the TODO and left in the read at the end of each setter.